src/VisualStudio/Next/Remote/RemoteHostService.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio.LanguageServices.Implementation.Remote;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Remote
+{
+    [ExportWorkspaceServiceFactory(typeof(IRemoteHostService)), Shared]
+    internal partial class RemoteHostServiceFactory : IWorkspaceServiceFactory
+    {
+        public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
+        {
+            return new RemoteHostService(workspaceServices.Workspace);
+        }
+
+        private class RemoteHostService : IRemoteHostService
+        {
+            private readonly Workspace _workspace;
+            private readonly SemaphoreSlim _lock;
+
+            private CancellationTokenSource _shutdown;
+            private Task<RemoteHost> _instance;
+
+            public RemoteHostService(Workspace workspace)
+            {
+                _workspace = workspace;
+                _lock = new SemaphoreSlim(initialCount: 1);
+            }
+
+            public void Enable()
+            {
+                using (_lock.DisposableWait())
+                {
+                    if (_instance != null)
+                    {
+                        // already enabled
+                        return;
+                    }
+
+                    // make sure we run it on background thread
+                    _shutdown = new CancellationTokenSource();
+                    _instance = Task.Run(() => StartInternalAsync(), _shutdown.Token);
+                }
+            }
+
+            public void Disable()
+            {
+                using (_lock.DisposableWait())
+                {
+                    if (_instance == null)
+                    {
+                        // already disabled
+                        return;
+                    }
+
+                    var instance = _instance;
+                    _instance = null;
+
+                    _shutdown.Cancel();
+
+                    try
+                    {
+                        instance.Result.Shutdown();
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        // _instance wasn't finished running yet.
+                    }
+                }
+            }
+
+            public async Task<RemoteHost> GetRemoteHostAsync(CancellationToken cancellationToken)
+            {
+                var instance = _instance;
```

Comment:
this needs to be wrapped in a lock to make sure htat the reads/writes don't get reordered.